src/cert_policy.rs

```diff
@@ -0,0 +1,1682 @@
+///
+/// > (e) If the subjectPublicKeyInfo field of ...
+///
+/// Does not apply.
+///
+/// > (f) Assign the certificate subjectPublicKey algorithm to ...
+///
+/// Does not apply.
+///
+/// > (g) If a name constraints extension is included in the certificate, ...
+///
+/// Does not apply.
+///
+/// > (h) If certificate i is not self-issued:
+///
+/// > (1) If explicit_policy is not 0, decrement explicit_policy by 1.
+///
+/// *explicit_policy* is virtually 0.
+///
+/// > (2) If policy_mapping is not 0, decrement policy_mapping by 1.
+///
+/// Policy mappings extension is not supported.
+///
+/// > (3) If inhibit_anyPolicy is not 0, decrement inhibit_anyPolicy by 1.
+///
+/// *inhibit_anyPolicy* is virtually always greater than 0, because
+/// *initial-any-policy-inhibit* is unset and inhibit anyPolicy extension is
+/// not supported.
+///
+/// > (i) If a policy constraints extension is included in the certificate, ...
+///
+/// Policy constraints extension is not supported.
+///
+/// > (j) If the inhibitAnyPolicy extension is included in the certificate, ...
+///
+/// Inhibit anyPolicy exxtension is not supported.
+///
+/// > (k) If certificate i is a version 3 certificate, ...
+///
+/// Does not apply.
+///
+/// > (l) If the certificate was not self-issued, verify that max_path_length ...
+///
+/// Does not apply.
+///
+/// > (m) If pathLenConstraint is present in the certificate ...
+///
+/// Does not apply.
+///
+/// > (n) If a key usage extension is present, ...
+///
+/// Does not apply.
+///
+/// > (o) Recognize and process any other critical extension ...
+///
+/// Does not apply.
+///
+/// #### 6.1.5. Wrap-Up Procedure
+///
+/// As detailed below, the steps described in this section do not apply.
+///
+/// > To complete the processing of the target certificate, perform the
+/// > following steps for certificate n:
+///
+/// > (a) If explicit_policy is not 0, decrement explicit_policy by 1.
+///
+/// *explicit_policy* is virtualy 0.
+///
+/// > (b) If a policy constraints extension is included in the certificate ...
+///
+/// Policy constraints extension is not supported.
+///
+/// > (c) Assign the certificate subjectPublicKey to working_public_key.
+///
+/// Does not apply.
+///
+/// > (d) If the subjectPublicKeyInfo field of the certificate ...
+///
+/// Does not apply.
+///
+/// > (e) Assign the certificate subjectPublicKey algorithm to ...
+///
+/// Does not apply.
+///
+/// > (f) Recognize and process any other critical extension ...
+///
+/// Does not apply.
+///
+/// > (g) Calculate the intersection of the valid_policy_tree and the
+/// > user-initial-policy-set, as follows:
+///
+/// Does not apply because we do not build the entire policy tree.
+/// Because we start from a specific policy OID in `user_initial_policy_set`,
+/// the valid policy tree path that our procedure finds is vitually included in
+/// the intersection of the *valid_policy_tree* and the
+/// *user-initial-policy-set*.
+pub fn validate_policy_tree_paths(
```

Comment:
I'm not sure what @cpu had in mind, but adding all of this code is definitely not what I had in mind when I was reading @cpu's suggestion to minimize the scope of the code added to this crate. Instead, I think we should keep your previous commit (that stores the policy extension) and make minimal extensions that exposes that information to downstream crates. Given the state as summarized previously, I'm not sure it makes sense for this crate's maintainers to maintain all this code.

Reply:
>  Instead, I think we should keep your previous commit (that stores the policy extension) and make minimal extensions that exposes that information to downstream crates.

That matches what I was thinking as well. Thanks for making it explicit :+1:

Reply:
The function `validate_policy_tree_path` has already been removed from this PR. `cert_policy.rs` is still huge, though.

`check_certificate_policies` function might be another concern. The other part is just the structural definition of Certificate Policies Extension and tests.

Reply:
I think our point is that we don't want any of this code to live in the rustls-webpki crate. As long as the caller is able to effectively get to a `&[u8]` representing the policy extension's contents, it seems like everything else should be able to live in other crates.

Reply:
I see. Now I understand my "previous commit" means the one I made in PR #211. But it would let certificates containing malformed certificate policies extensions marked as critical pass. I think [that was the reason we rejected PR#211](https://github.com/rustls/webpki/pull/211#issuecomment-1827358947).

Reply:
Okay, that's fair. @ctz what do you think, should we:

1. Accept parsing code like in this PR to ensure we don't ignore invalid policies
2. Continue to not process/remember policies at all
3. Some middle ground where we do high-level parsing of the policy extension, but not in too much depth
4. Some other solution, perhaps involving configuration for another optional callback/trait

Reply:
My concern is that we continue to reject critical extensions if we don't fully and unconditionally implement the constraints they specify. That is what critical 

So solutions acceptable to me would bind together the acceptance of critical extensions with the validation and enforcement of their contents.

But I think a better overall solution is to use one of the more general certificate verification crates, rather than extending one that is designed to be specific to the webpki.

Reply:
> But I think a better overall solution is to use one of the more general certificate verification crates, rather than extending one that is designed to be specific to the webpki.

@ctz Do you have any recommended crates to look into? I briefly searched for some crates with keywords like `x509` and `pkix`:
- [`x509-parser`](https://docs.rs/x509-parser/latest/x509_parser/):
  - does not seem to expose any method to specifically parse certificate policies extensions
  - lacks types for policy qualifiers
- [`x509-cert`](https://docs.rs/x509-cert/latest/x509_cert/index.html):
  - seems to expose the way to parse most of the necessary types but lacks `DisplayText` variants for `VisibleString` and `BMPString` which I faced in the test cases I added
- [`pkix`](https://docs.rs/pkix/latest/pkix/index.html):
  - seems to expose the way to parse all the necessary types
- [`rasn-pkix`](https://docs.rs/rasn-pkix/latest/rasn_pkix/):
  - similar to `pkix`